### DISCUSSION TITLE
How to apply/implement filtering on child resolvers?

### DISCUSSION THREAD

**Discussion by gabbigum** [Netflix/dgs-framework#560] [Q&A]
Following the schema from Netflix DGS Framework documentation, how can we get all the shows that have reviews with star rating 3? More generalised, how can we apply filter on child resolvers? Thanks in advance!

```
type Query {
  shows: [Show]
}

type Show {
  # The showId may or may not be there, depending on the scenario.
  showId: ID
  title: String
  reviews: [Review]
}

type Review {
  starRating: Int
}
```

**Comment by paulbakker** [comment]
In the schema you can add an argument to a field. E.g. here I added a `minumumStarRating` to `reviews`.

```
type Query {
  shows: [Show]
}

type Show {
  # The showId may or may not be there, depending on the scenario.
  showId: ID
  title: String
  reviews(minimumStarRating: Int): [Review]
}

type Review {
  starRating: Int
}
```

In the datafetcher you can get the argument with a `@InputArgument`:

```
public List<Review> reviews(@InputArgument Integer minimumStarRating) {
    // Load and filter the reviews using the argument.
}
```

**Reply by gabbigum** [reply]
Yes, I've applied the filter like this but I guess I didn't asked the question correctly. My question is how can we really filter the response? As a user I expect to receive **ONLY** the shows with minimumStarRating. Currently my request/response looks like this:

```
# Request
shows {
    title
    reviews(minimumStarRating: 3) {
      starRating
    }
  }
```

```
# Response
"shows": [
      {
        "title": "Stranger Things",
        "reviews": [
          {
            "starRating": 5
          },
          {
            "starRating": 5
          },
          {
            "starRating": 4
          }
        ]
      },
      {
        "title": "Ozark", # This shouldn't be here 
        "reviews": null # because it doesn't meet my filtering criteria
      },
      {
        "title": "The Crown",
        "reviews": [
          {
            "starRating": 3
          },
          {
            "starRating": 4
          },
          {
            "starRating": 4
          }
        ]
      }
]
```

**Reply by paulbakker** [reply]
I would argue that your schema design is not optimal in that case. The rating filter should be at the `shows` level, not the `reviews` level, if you want to use it to filter the shows.

If you put it on the level of the reviews, I would expect it to only filter the reviews.

That being said, you could get it to work by using a "look ahead", which is basically looking at what fields are selected inside the shows datafetcher. This mechanism is explained here: https://netflix.github.io/dgs/advanced/context-passing/#pre-loading
I would re-think the schema first though.

**Reply by gabbigum** [reply]
Yeah, I know that the schema design looks off and that our use case might seem strange. It is just that our entities are linked by foreign keys (strings with resource links). That's why performing the filtration in the query resolver would be too expensive of an operation when we have millions of entries in our database. 

The solution that we think of is if we can somehow intercept the query result as it is getting built and modify it according to our filtering criteria. I guess that's more related to the internals of graphql-java than to DGS, but any advice on doing that would be of great help.

Our engineering team from VMware thanks you for the quick response and feedback!

**Reply by paulbakker** [reply]
That sounds exactly like what's described in the docs I linked: https://netflix.github.io/dgs/advanced/context-passing/#pre-loading.

Basically in the parent resolver you inspect what fields are requested deeper down, and craft a (database) query based on that.

**Reply by asavov** [reply]
Hi Paul,

My name is Alex, dev at VMware. @gabbigum and I are working on adopting DGS framework into one of our Multi-Cloud products.

Thanks for prompt response pointing to `pre-load` approach.

IMO this approach might work well in case of:
1) few child entities and
2) **direct** access to DB (which is not an option for us)

Still it eliminates the need for Resolvers (respectively @DgsData counterparts), thus limiting the **extensibility** through Resolvers and converting QueryResolver to God-Object (which is aware of **all children** and how to resolve them).

What are your thoughts about implementing following **INNER JOIN-like** approach?

- Framework calls root/query resolver and gets result of 5 Shows.
- For each Show the framework calls child `review` resolver with review-related filter (let's say reviews with 3+ starts and author name Paul).
- For Show 3 and 5 the review-resolver returns empty/null reviews, due to not meeting the filtering criteria.
- On `merge-reviews-into-show` phase the Framework skips/ignores/excludes those Shows for which reviews are null/empty.
- As a result only Show 1, 2, 4 are returned to the client.

Questions:
- Is above approach **aligned** with DGS framework **principles** (which seems to follow **LEFT OUTER JOIN-like** approach) and GraphQL **ideology**?
- Can it be **provided** by DGS itself? Or can it be achieved through writing **custom** `Instrumentation` XXX APIs (or any other hook provided by the framework)?

Thanks for helping VMware adopt DGS!

-Alex Savov

**Reply by 1zg12** [reply]
I think this is a valid use case. If this feature is implemented, it could help a lot developers.

Our similar use case here:

```
type Query {
  position: [Position]
}

type Position{
  date: String
  hierarchyId: String
  securityName: String
   ...
  hierarchy: Hierarchy
}

type Hierarchy{
  portfolio: String
  book: String
  ...
}
```

image when a user query the data, the user could only interested in his/her only positions, something like 
```
{
 position (date: String, portfolio: String){
   date
   securityName
   hierarchy{
     portfolio
     book
   }
 }
}
```

I guess ultimately, it's how the backend is handling the relationship (edges). At the moment, dgs/graphql-java is behaving like `left-outer-join` (in RDBMS terms), but really there should be an option for user to pick `inner-join` instead. I think that's the feature missing here.

I guess something like
```
@DgsData(parentType="Position", field="hierarchy", `join=InnerJoin`)
public Completable<Future<Hierarchy>> batchloadHierarchy(DataFetchingEnvironment dfe){
   //only return the hierarchy belong to the portfolio
}
```
then when dgs/graphql-jar stitching the position and hierarchy data, if it's InnerJoin, it should drop the position without hierarchy.

**Comment by asavov** [comment]
@paulbakker Any update/response/hint/recommendation is still highly appreciated. Thanks!

**Comment by paulbakker** [comment]
Filtering a top level object (`shows`) based on a field (`reviews`) just seem fundamentally a broken schema design. Returning null should definitely not have such as side effect.
Also note that that's implemented in `graphql-java` not in the DGS Framework.

I'm not sure what the best way is to deal with this, other than the look ahead mechanism I mentioned. Again, the schema design isn't correct, so that's what really should be fixed.

I understand that this sounds incredibly unhelpful, but there really isn't anything the framework can or should do here.

**Comment by jamesmbowler** [comment]
I needed something similar, this is what I've got so far: https://github.com/jamesmbowler/robertgraphql
All joins must be many/many because hibernate lazy loading issues. I wanted the graphql request to create one sql query. It's really raw right now but I am improving it.
